docs/spec/reactors/consensus/proposer-selection.md

```diff
@@ -2,45 +2,268 @@
+#### Voting Power Change
+Consider again the earlier example and assume that the voting power of p1 is changed to 4:
+
+
+Validator | p1| p2 
+----------|---| ---
+VP        | 4 | 3
+
+Let's also assume that at the last round R at height H before this change the proposer priorites were as shown in first row (H/R end). As it can be seen, the procedure could continue with round 1 at next height without changes as before. 
+
+|Round Priority| -2 | -1 | 0    | 1   | 2   | 3 | 4 | 5 | Comment
+|--------------| ---|--- |------|---  |---  |---|---|---|--------
+| H/R end      |    | p2 |      |  p1 |     |   |   |   |update VP(p1)
+| H+1/0        |    |    |      |     | p2  |   |   | p1|A(i)+=VP(i)
+|              | p1 |    |      |     | p2  |   |   |   |A(p1)-= P
+
+However, when a validator changes power from a high to a low value it may be stuck at the end of the queue for a long time if it happened to be there (e.g. it just proposed). This scenario is considered again after the New Validator section.
+
+#### Validator Removal
+Consider a new example with set.
+
+Validator | p1 | p2 | p3 |
+--------- |--- |--- |--- |
+VP        | 1  | 2  | 3  |
+
+Let's assume that the last round at height H was R and the proposer priorities were as shown in first row (H/R end) with their sum being 0. At this point p2 is removed and round 1 of next height H+1 runs. At the end of the round (penultimate row) the sum of priorities is -2 (minus the priority of the removed process). 
+ 
+
+|Round Priority |-3 | -2 | -1 | 0  | 1   | 2   | 3 | 4 | 5 | Comment
+|---------------|-- | ---|--- |--- |---  |---  |---|---|---|--------
+| H/R end       |p3 |    |    |    | p1  | p2  |   |   |   |remove p2
+| H+1/0         |   |    |    | p3 |     | p1  |   |   |   |A(i)+=VP(i)
+|               |   | p1 |    | p3 |     |     |   |   |   |A(p1)-= P
+| *new step*    |   |    | p1 |    | p3  |     |   |   |   |A(i) -= avg
+
+The procedure could continue without modifications. However, it is possible that after a sufficiently large number of modifications in validator set, the priority values would migrate towards maximum or minimum allowed values causing truncations due to overflow detection.
+For this reason, the selection procedure adds another step (see last row) that shifts the current priority values left or right such that the average remains close to 0. Due to integer division the priority sum is in {-1, 0, 1}
+
+The modified selection algorithm is:
+
+    def singleProposerSelection (vset):
+        for each validator i in vset:
+            A(i) += VP(i)
+        prop = max(A)
+        A(prop) -= P
+    
+    def ProposerSelection (vset, times):
+        for in in range(times):
+            singleProposerSelection(vset)
+
+        // shift priorities with the average
+        avg = sum(A(i) for i in vset)/len(vset)
+        for each validator i in vset:
+            A(i) -= avg
+
+Note that the shifting operation happens after singleProposerSelection() has run times times.
```

Comment:
As already discussed, the shifting operation should probably happen only after validator set is modified so that sum of priorities is always around 0 before round is started. This can be addressed together with the corresponding code modifications is a separate PR.

Reply:
Ok, let's do this once we wrap up the current spec.